src/coap_openssl.c

```diff
@@ -1132,24 +1361,103 @@ missing_ENGINE_load_cert(const char *cert_id) {
     params.cert = NULL;
   }
   return params.cert;
 }
 
 #if OPENSSL_VERSION_NUMBER < 0x10101000L && COAP_SERVER_SUPPORT
+
+/* OpenSSL < 1.1.1 and Server */
+static int
+install_engine_public_cert_ctx(ENGINE *engine, SSL_CTX *ctx, const char *public_cert) {
+  X509 *x509;
+
+  x509 = missing_ENGINE_load_cert(engine, public_cert);
+  if (!x509) {
+    coap_log_warn("*** setup_pki: (D)TLS: %s: Unable to load "
+                  "Server Certificate\n",
+                  public_cert);
+    return 0;
+  }
+  if (!SSL_CTX_use_certificate(ctx, x509)) {
+    coap_log_warn("*** setup_pki: (D)TLS: %s: Unable to configure "
+                  "Server Certificate\n",
+                  public_cert);
+    X509_free(x509);
+    return 0;
+  }
+  X509_free(x509);
+  return 1;
+}
+
+/* OpenSSL < 1.1.1 and Server */
+static int
+install_engine_private_key_ctx(ENGINE *engine, SSL_CTX *ctx, const char *private_key) {
+  EVP_PKEY *pkey = ENGINE_load_private_key(engine,
+                                           private_key,
+                                           NULL, NULL);
+
+  if (!pkey) {
+    coap_log_warn("*** setup_pki: (D)TLS: %s: Unable to load "
+                  "Server Private Key\n",
+                  private_key);
+    return 0;
+  }
+  if (!SSL_CTX_use_PrivateKey(ctx, pkey)) {
+    coap_log_warn("*** setup_pki: (D)TLS: %s: Unable to configure "
+                  "Server Private Key\n",
+                  private_key);
+    EVP_PKEY_free(pkey);
+    return 0;
+  }
+  EVP_PKEY_free(pkey);
+  return 1;
+}
+
+/* OpenSSL < 1.1.1 and Server */
+static int
+install_engine_ca_ctx(ENGINE *engine, SSL_CTX *ctx, const char *ca) {
+  X509 *x509;
+  X509_STORE *st;
+
+  x509 = missing_ENGINE_load_cert(engine,
+                                  ca);
+  if (!x509) {
+    coap_log_warn("*** setup_pki: (D)TLS: %s: Unable to load "
+                  "Server CA Certificate\n",
+                  ca);
+    return 0;
+  }
+  if (!SSL_CTX_add_client_CA(ctx, x509)) {
+    coap_log_warn("*** setup_pki: (D)TLS: %s: Unable to configure "
+                  "Server CA Certificate\n",
+                  ca);
+    X509_free(x509);
+    return 0;
+  }
+  st = SSL_CTX_get_cert_store(ctx);
+  add_ca_to_cert_store(st, x509);
+  X509_free(x509);
+  return 1;
+}
+
+/* OpenSSL < 1.1.1 and Server */
 static int
 setup_pki_server(SSL_CTX *ctx,
-                 const coap_dtls_pki_t *setup_data
-                ) {
+                 const coap_dtls_pki_t *setup_data) {
   switch (setup_data->pki_key.key_type) {
+  /* OpenSSL < 1.1.1 and Server */
   case COAP_PKI_KEY_PEM:
     if (setup_data->pki_key.key.pem.public_cert &&
         setup_data->pki_key.key.pem.public_cert[0]) {
-      if (!(SSL_CTX_use_certificate_file(ctx,
-                                         setup_data->pki_key.key.pem.public_cert,
-                                         SSL_FILETYPE_PEM))) {
+      if (defined_engine && install_engine_public_cert_ctx(defined_engine, ctx,
```

Comment:
can we have a config param  setup_data->pki_key.key.pem.public_cert_file check rather that making it mandate to have a certificate_id(string) in case of engine defined?

Reply:
Would something more generic like a new `case COAP_PKI_KEY_ENGINE` where you define the ENGINE information along the lines of
````
/**
 * The structure to define the format of the engine parameter definition */
 */
typedef struct coap_engine_type_t {
  uint32_t pem:1; /** Engine entry type Use as file PEM */
  uint32_t pem_buf:1; /** Engine entry type Use as buffer PEM */
  uint32_t der:1; /** Engine entry type Use as file DER */
  uint32_t engine:1; /** Engine entry type Use Engine */
} coap_engine_type_t;

/**
 * The structure that holds the PKI ENGINE definitions.
 */
typedef struct coap_pki_key_engine_t {
  const char *ca;            /**< engine: Common CA Certificate */
  const char *public_cert;   /**< engine: Public Cert */
  const char *private_key;   /**< engine: Private Key */
  coap_engine_type_t ca_type; /**< engine: Common CA type */
  coap_engine_type_t public_cert_type; /**< engine: Public Cert type */
  coap_engine_type_t pri_key_t_type;   /**< engine: Private Key type */
} coap_pki_key_engine_t;
````
work for you?

Thus if one or more of `ca`, `public_cert `and `private_key `are defined as type `engine`, the engine would only get called for the appropriate entry (assuming the engine has been configured).  Otherwise, they would not use the engine and call the appropriate function as per the other case COAP_PKI_KEY__* statements

I guess that this is all down to the specification of the ENGINE interface and what functionality is supported / needed here.

Reply:
I guess we can use existing enum for coap_engine_type instead struct:
/**
 * The enum used for determining the PKI key formats.
 */
typedef enum coap_pki_key_t {
  COAP_PKI_KEY_PEM = 0,        /**< The PKI key type is PEM file */
  **COAP_PKI_KEY_ASN1**,           /**< The PKI key type is ASN.1 (DER) buffer */
  COAP_PKI_KEY_PEM_BUF,        /**< The PKI key type is PEM buffer */
  **COAP_PKI_KEY_PKCS11**,         /**< The PKI key type is PKCS11 (DER) */
} coap_pki_key_t;

This makes it easier to define and compare the types of the CA, public certificate, and private key.

Reply:
Thanks for this - I had realized about the enum after I had stopped working for the day.

Reply:
> Would something more generic like a new `case COAP_PKI_KEY_ENGINE` where you define the ENGINE information along the lines of
> 
> ```
> /**
>  * The structure to define the format of the engine parameter definition */
>  */
> typedef struct coap_engine_type_t {
>   uint32_t pem:1; /** Engine entry type Use as file PEM */
>   uint32_t pem_buf:1; /** Engine entry type Use as buffer PEM */
>   uint32_t der:1; /** Engine entry type Use as file DER */
>   uint32_t engine:1; /** Engine entry type Use Engine */
> } coap_engine_type_t;
> 
> /**
>  * The structure that holds the PKI ENGINE definitions.
>  */
> typedef struct coap_pki_key_engine_t {
>   const char *ca;            /**< engine: Common CA Certificate */
>   const char *public_cert;   /**< engine: Public Cert */
>   const char *private_key;   /**< engine: Private Key */
>   coap_engine_type_t ca_type; /**< engine: Common CA type */
>   coap_engine_type_t public_cert_type; /**< engine: Public Cert type */
>   coap_engine_type_t pri_key_t_type;   /**< engine: Private Key type */
> } coap_pki_key_engine_t;
> ```
> 
> work for you?
> 
> Thus if one or more of `ca`, `public_cert `and `private_key `are defined as type `engine`, the engine would only get called for the appropriate entry (assuming the engine has been configured). Otherwise, they would not use the engine and call the appropriate function as per the other case COAP_PKI_KEY__* statements
> 
> I guess that this is all down to the specification of the ENGINE interface and what functionality is supported / needed here.

are you going to implement this soon?

Reply:
Just pushed a code update where there is a new `COAP_PKI_KEY_DEFINE `and `coap_pki_key_define_t `that gives the required flexibility.

Please test and comment on. An example of usage is in `coap_tls_library(3).`

There still is some code tidy up to do.  `COAP_PKI_KEY_DEFINE `is currently only supported for OpenSSL and TinyDTLS.

Reply:
Done testing with the latest changes. It is working fine with this generic approach which gives more flexibility to the developer. 

Appreciate your quick response and fix.

Reply:
Thanks for testing this out.  Will carry on with some code tidy up.